src/viewport-observer.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {devAssert} from '../src/log';
+import {isIframed} from './dom';
+
+/**
+ * Returns an IntersectionObserver tracking the Viewport.
+ * Only use this if x-origin iframe rootMargin support is considered nice-to-have,
+ * since if not supported this InOb will fallback to one without rootMargin.
+ *
+ * - If iframed: rootMargin is ignored unless natively supported (Chrome 81+).
+ * - If not iframed: all features work properly in both polyfill and built-in.
+ *
+ * @param {function(!Array<!IntersectionObserverEntry>)} ioCallback
+ * @param {!Window} win
+ * @param {number=|!Array<number>=} threshold
+ *
+ * @return {!IntersectionObserver}
+ */
+export function createViewportObserver(ioCallback, win, threshold) {
```

Comment:
Is threshold support coming later? Nothing seems to use it.

Reply:
Video player inViewport detection uses it: https://github.com/ampproject/amphtml/pull/30647/files#diff-c19d7c3d268d0eda98db54f0cfbb91747f92cbfba8f466e753e410a204f1c8f8R202

Reply:
not happy about the fn name length, but I can't argue with the increase in clarity.  will update now